### PR TITLE
adds ex_security_group_ids to ec2 in order to be able to launch nodes with security groups on a VPC

### PR DESCRIPTION
**Pull request by Itxaka**
Due to amazon having a different param for the security groups on an instance launched on a VPC and libcloud already supports VPCs (ex_subnet) this is the piece is missing to support the security groups on a VPC node
- adds a ex_security_group_ids kwarg which is a list of security group ids
- checks for the existence of ex_subnet if ex_security_group_ids is called, as it depends on it to make sense
- adds a test to see if the params are being constructed correctly

EC2 API page: http://docs.aws.amazon.com/AWSEC2/latest/APIReference/ApiReference-query-RunInstances.html

```
SecurityGroupId.n
One or more security group IDs. You can create a security group using CreateSecurityGroup.
Type: String
Default: Amazon EC2 uses the default security group.
Required: No
```

```
SecurityGroup.n
[EC2-Classic, default VPC] One or more security group names. For a nondefault VPC, you must use SecurityGroupId.n.
Type: String
Default: Amazon EC2 uses the default security group.
Required: No
```

### CODE REVIEW

**Comment by Itxaka** [comment]
@Kami :) :shipit:

File: libcloud/compute/drivers/ec2.py [hunk 1]
```diff
@@ -2187,6 +2191,20 @@ def create_node(self, **kwargs):
                 params['SecurityGroup.%d' % (sig + 1,)] =\
                     security_groups[sig]
 
+        if 'ex_security_group_ids' in kwargs and 'ex_subnet' not in kwargs:
+            raise ValueError('You can only supply ex_security_group_id'
+                             ' combinated with ex_subnet')
+
+        security_group_id = kwargs.get('ex_security_group_ids', None)
```
Comment (Kami) [inline libcloud/compute/drivers/ec2.py]: This variable stores a list so it should be called `security_group_ids`.

File: libcloud/test/compute/test_ec2.py [hunk 2]
```diff
@@ -871,6 +872,22 @@ def test_create_node_ex_security_groups(self):
                           ex_securitygroup=security_groups,
                           ex_security_groups=security_groups)
 
+    def test_create_node_ex_security_group_id(self):
```
Comment (Kami) [inline libcloud/test/compute/test_ec2.py]: Also needs a test case for edge case (when user specifies `ex_security_group_ids`, but not `ex_subnet`).

File: libcloud/compute/drivers/ec2.py [hunk 3]
```diff
@@ -2135,6 +2135,10 @@ def create_node(self, **kwargs):
                                         assign to the node.
         :type       ex_security_groups:   ``list``
 
+        :keyword    ex_security_group_ids: A list of ids of security groups to
```
Comment (Kami) [inline libcloud/compute/drivers/ec2.py:2138]: Edit: Ignore this comment - pr description already addresses it.

It's just unfortunate that existing argument is called `ex_security_groups` instead of `ex_security_group_names`.

File: libcloud/test/compute/test_ec2.py (hunk 2 above)
Comment (Itxaka) [inline libcloud/test/compute/test_ec2.py]: I tried but could not catch the Exception with self.assertRaises, it actually raised the Exception. Will look into this.

Comment (Kami) [inline libcloud/test/compute/test_ec2.py]: That's probably because you were actually calling the method yourself and not just passing a callable to `assertRaises` (you should only pass callable + arguments to assertRaises and let assetRaises call the callable).

For example:

``` python
self.assertRaises(ValueError, myfunc(blah, foo=bar))
```

Not good because you are calling the function yourself.

For example:

``` python
self.assertRaises(ValueError, myfunc, blah, foo=bar)
```

Good - you are letting assertRaises call the function and intercept the exceptions.

**Comment by Itxaka** [comment]
@Kami ?

**Comment by Kami** [comment]
@Itxaka Sorry for the delay.

I will have a look later today and hopefully also get it merged then.

**Comment by Kami** [comment]
LGTM. Patch has been merged into trunk.

Thanks.
